### PR TITLE
Timezone support for timers

### PR DESCRIPTION
**Pull request by zerkms**
This PR adds real support for timezones in `OnCalendar` parameter of timers.

Timezones from the output of `timedatectl list-timezones` are supported.

Timezone component comes as a very latest part of the specification, after the time.

Examples:

```
2015-10-25 01:00:00 Asia/Vladivostok
weekly Pacific/Auckland
```

### CODE REVIEW

File: src/basic/calendarspec.c [hunk 1]
```diff
@@ -876,6 +884,53 @@ static int parse_calendar_time(const char **p, CalendarSpec *c) {
         return r;
 }
 
+static int is_valid_timezone(char *tz, bool *result)
+{
```
Comment (poettering) [inline src/basic/calendarspec.c]: nitpick, please follow coding style: opening bracket for functions on same line as function name please

File: src/basic/calendarspec.c [hunk 2]
```diff
@@ -876,6 +884,53 @@ static int parse_calendar_time(const char **p, CalendarSpec *c) {
...
+        pid_t pid;
+
+        if (tz == NULL || strlen(tz) == 0) {
```
Comment (poettering) [inline src/basic/calendarspec.c]: use isempty()

File: src/basic/calendarspec.c [hunk 3]
```diff
@@ -876,6 +884,53 @@ static int parse_calendar_time(const char **p, CalendarSpec *c) {
         return r;
 }
 
+static int is_valid_timezone(char *tz, bool *result)
```
Comment (poettering) [inline src/basic/calendarspec.c]: there's already timezone_is_valid() defined in src/basic/time-util.c. Please use that.

File: src/basic/calendarspec.c [hunk 4]
```diff
@@ -876,6 +884,53 @@ static int parse_calendar_time(const char **p, CalendarSpec *c) {
...
+{
+        pid_t pid;
+
+        if (tz == NULL || strlen(tz) == 0) {
+                *result = false;
+                return 0;
+        }
+
+        bool *shared = mmap(NULL, sizeof *shared, PROT_READ|PROT_WRITE, MAP_SHARED|MAP_ANONYMOUS, -1, 0);
```
Comment (poettering) [inline src/basic/calendarspec.c]: nitpick: see CODING_STYLE, we avoid defining variables in the middle of functions, and we avoid mixing variable declarations and function invocations in one line

File: src/basic/calendarspec.c [hunk 5]
```diff
@@ -876,6 +884,53 @@ static int parse_calendar_time(const char **p, CalendarSpec *c) {
...
+
+        if (shared == MAP_FAILED) {
+                return log_error_errno(errno, "mmap() failed: %m");
+        }
```
Comment (poettering) [inline src/basic/calendarspec.c]: nitpick: see CODING_STYLE, not {} around single-line if blocks

File: src/basic/calendarspec.c [hunk 6]
```diff
@@ -876,6 +884,53 @@ static int parse_calendar_time(const char **p, CalendarSpec *c) {
...
+
+        pid = fork();
+
+        if (pid == -1) {
+                munmap(shared, sizeof *shared);
```
Comment (poettering) [inline src/basic/calendarspec.c]: while we don#t do that for much of the older code, for new code please explicitly cast the return value of syscalls whose return value you want to ignore to (void), so that static analyzers understand this

File: src/basic/calendarspec.c [hunk 7]
```diff
@@ -876,6 +884,53 @@ static int parse_calendar_time(const char **p, CalendarSpec *c) {
...
+                return log_error_errno(errno, "Failed to fork: %m");
+        }
+
+        if (pid == 0) {
+                setenv("TZ", tz, 1);
```
Comment (poettering) [inline src/basic/calendarspec.c]: please check for setenv() failing

File: src/basic/calendarspec.c [hunk 8]
```diff
@@ -876,6 +884,53 @@ static int parse_calendar_time(const char **p, CalendarSpec *c) {
...
+                tzset();
+
+                *shared = strlen(tzname[1]) > 0;
+
+                _exit(0);
```
Comment (poettering) [inline src/basic/calendarspec.c]: not that it matters, but `_exit(EXIT_SUCCESS)` is prettier

File: src/basic/calendarspec.c [hunk 9]
```diff
@@ -876,6 +884,53 @@ static int parse_calendar_time(const char **p, CalendarSpec *c) {
...
+        if (pid == -1) {
+                munmap(shared, sizeof *shared);
+                return log_error_errno(errno, "Failed to fork: %m");
+        }
+
+        if (pid == 0) {
+                setenv("TZ", tz, 1);
+                tzset();
+
+                *shared = strlen(tzname[1]) > 0;
```
Comment (poettering) [inline src/basic/calendarspec.c]: maybe !isempty()?

File: src/basic/calendarspec.c [hunk 10]
```diff
@@ -876,6 +884,53 @@ static int parse_calendar_time(const char **p, CalendarSpec *c) {
...
+
+                _exit(0);
+        }
+
+        if (waitpid(pid, NULL, 0) == -1) {
+                int waitpid_errno = errno;
+                munmap(shared, sizeof *shared);
+                return log_error_errno(waitpid_errno, "waitpid() failed: %m");;
```
Comment (poettering) [inline src/basic/calendarspec.c]: consider using wait_for_terminate()?

File: src/basic/calendarspec.c [hunk 11]
```diff
@@ -876,6 +884,53 @@ static int parse_calendar_time(const char **p, CalendarSpec *c) {
...
+        }
+
+        bool tmp_result = *shared;
+        if (munmap(shared, sizeof *shared) != 0) {
+                return log_error_errno(errno, "munmap() failed: %m");;
+        }
+
+        *result = tmp_result;
```
Comment (poettering) [inline src/basic/calendarspec.c]: btw, using the mmap() thing appears unnecessary, for this the exit code of the child should be entirely enough for the information to transport

File: src/basic/calendarspec.c [hunk 12]
```diff
@@ -919,6 +975,29 @@ int calendar_spec_from_string(const char *p, CalendarSpec **spec) {
...
+                                }
+
+                                if (timezone_is_valid) {
+                                        c->timezone = malloc(strlen(timezone) + 1);
+                                        if (c->timezone == NULL) {
+                                                r = -EINVAL;
+                                                goto fail;
+                                        }
+
+                                        strcpy(c->timezone, timezone);
```
Comment (poettering) [inline src/basic/calendarspec.c]: use strdup() for this

**Comment by poettering** [comment]
btw, the language for calendar timers and for timestamps are supposed to stay similar, i.e. that each timestamp specification is also a valid calendar timer specification (though one which will fire only once). Hence, parse_timestamp() should really be updated the same way at the same time

File: src/basic/calendarspec.c (hunk 3 above)
Comment (zerkms) [reply src/basic/calendarspec.c]: @poettering yep, it was removed 2 days ago, not sure how you ended up commenting the obsolete commit. Apologise for not squashing it first.

Comment (poettering) [reply src/basic/calendarspec.c]: Ah, please always push "perfect" PRs. I.e. push PRs that do not reflect your development history but rather reflect logical steps, that are easy to review. Or to say this differently: we want each commit to be nicely bisectable.

**Comment by poettering** [comment]
btw, we prefer merging PRs in "perfect" form, i.e. that individual commits of a single PR are easily-reviewable logical steps but do not necessarily reflect development history. Or in other words: each commit should be fully bisectable.

**Comment by zerkms** [comment]
@poettering yep, I remember that. I will squash all when I'm done with your requests: I have completed styling and minor stuff and now it's `parse_timestamp` that is left. It's slightly more complicated so I believe it would take couple more days. After that I will make squash for all the commits and it will only be the single commit rebased to the latest `master`.

**Comment by bluetech** [comment]
Thanks for doing this, it's a very needed feature.

Huh, doing a full fork just to fool glibc; that's a clever hack I guess :) I think it deserves a comment though, explaining why the entire dance is needed, and warning that the function is "heavy" (so don't call it in a loop).

**Comment by zerkms** [comment]
@poettering @bluetech 

I think I'm done.

Now timezones are supported in both systemd.timer OnCalendar and timedatectl

File: src/basic/calendarspec.c [hunk 13]
```diff
@@ -349,9 +352,7 @@ int calendar_spec_to_string(const CalendarSpec *c, char **p) {
         fputc_unlocked(':', f);
         format_chain(f, 2, c->microsecond, true);
 
-        if (c->utc)
-                fputs_unlocked(" UTC", f);
-        else if (IN_SET(c->dst, 0, 1)) {
+        if (IN_SET(c->dst, 0, 1)) {
```
Comment (poettering) [inline src/basic/calendarspec.c]: i am surprised about the ordering here, shouldn't the explicit timezone if it is configured be shown instead of the implicit one here?

Also, I think it would make sense to continue to support UTC explicitly. It is after all a very special timezone, probably pretty regularly used and the only one besides the local one which we can support without doing the fork() magic. Hence, let's support calendar in three different modes: in the local timezone, in UTC and in any other timezone.

File: src/basic/calendarspec.c [hunk 14]
```diff
@@ -595,7 +601,7 @@ static int calendarspec_from_time_t(CalendarSpec *c, time_t time) {
         if (r < 0)
                 return r;
 
-        c->utc = true;
+        c->timezone = strdup("UTC");
```
Comment (poettering) [inline src/basic/calendarspec.c]: misses OOM check (see CODING_STYLE). That said, I think we should keep explicit UTC support in place, see comment above

File: src/basic/calendarspec.c [hunk 15]
```diff
@@ -888,37 +894,46 @@ int calendar_spec_from_string(const char *p, CalendarSpec **spec) {
...
+                if (last_space != NULL) {
+                        const char *timezone = last_space + 1;
 
-                /* Found one of the two timezones specified? */
-                if (IN_SET(j, 0, 1)) {
-                        p = strndupa(p, e - p - 1);
-                        c->dst = j;
+                        if (strcasecmp(timezone, "UTC") == 0) {
+                                timezone = "UTC";
+                        }
```
Comment (poettering) [inline src/basic/calendarspec.c]: no, {} around single-line if blocks. See CODING_STYLE.

File: src/basic/calendarspec.c [hunk 16]
```diff
@@ -888,37 +894,46 @@ int calendar_spec_from_string(const char *p, CalendarSpec **spec) {
...
+                last_space = strrchr(p, ' ');
+
+                if (last_space != NULL) {
+                        const char *timezone = last_space + 1;
 
-                /* Found one of the two timezones specified? */
-                if (IN_SET(j, 0, 1)) {
-                        p = strndupa(p, e - p - 1);
-                        c->dst = j;
+                        if (strcasecmp(timezone, "UTC") == 0) {
```
Comment (poettering) [inline src/basic/calendarspec.c]: please use strcaseeq()

File: src/basic/calendarspec.c [hunk 17]
```diff
@@ -888,37 +894,46 @@ int calendar_spec_from_string(const char *p, CalendarSpec **spec) {
...
+                                timezone = "UTC";
+                        }
+
+                        if (timezone_is_valid(timezone)) {
+                                c->timezone = strdup(timezone);
```
Comment (poettering) [inline src/basic/calendarspec.c]: missing oom check

File: src/basic/calendarspec.c [hunk 18]
```diff
@@ -1307,17 +1322,72 @@ int calendar_spec_next_usec(const CalendarSpec *spec, usec_t usec, usec_t *next)
...
+
+                shared->return_value = calendar_spec_next_usec_impl(spec, usec, &shared->next);
+
+                _exit(EXIT_SUCCESS);
+        }
+
+        r = wait_for_terminate(pid, NULL);
+        if (r < 0) {
+                (void) munmap(shared, sizeof *shared);
+                return log_error_errno(r, "waitpid() failed: %m");
```
Comment (poettering) [inline src/basic/calendarspec.c]: so this is a library function, and hence shouldn't log on its own above LOG_DEBUG. Let the caller do that. See CODING_STYLE about that

**Comment by zerkms** [comment]
@poettering sorry for those mistakes.

I agree keeping UTC as a separate implementation makes sense.

I also fixed all other relevant issues.

File: src/basic/calendarspec.c [hunk 19]
```diff
@@ -919,6 +926,23 @@ int calendar_spec_from_string(const char *p, CalendarSpec **spec) {
...
+
+                        last_space = strrchr(p, ' ');
+
+                        if (last_space != NULL) {
+                                char *timezone = last_space + 1;
+
+                                if (timezone_is_valid(timezone)) {
+                                        c->timezone = strdup(timezone);
+                                        if (!c->timezone) {
+                                                r = -EINVAL;
```
Comment (keszybz) [inline src/basic/calendarspec.c]: ENOMEM, not EINVAL.

File: src/basic/calendarspec.c [hunk 20]
```diff
@@ -919,6 +926,23 @@ int calendar_spec_from_string(const char *p, CalendarSpec **spec) {
                 if (IN_SET(j, 0, 1)) {
                         p = strndupa(p, e - p - 1);
                         c->dst = j;
+                } else {
+                        char *last_space;
```
Comment (keszybz) [inline src/basic/calendarspec.c]: `const char *`.

File: src/basic/calendarspec.c [hunk 21]
```diff
@@ -919,6 +926,23 @@ int calendar_spec_from_string(const char *p, CalendarSpec **spec) {
...
+
+                        last_space = strrchr(p, ' ');
+
```
Comment (keszybz) [inline src/basic/calendarspec.c]: This empty line is unnecessary.

File: src/basic/calendarspec.c [hunk 22]
```diff
@@ -919,6 +926,23 @@ int calendar_spec_from_string(const char *p, CalendarSpec **spec) {
...
+                        if (last_space != NULL) {
+                                char *timezone = last_space + 1;
```
Comment (keszybz) [inline src/basic/calendarspec.c]: This should be `const char *` too.

File: src/basic/time-util.c [hunk 23]
```diff
@@ -874,6 +879,71 @@ int parse_timestamp(const char *t, usec_t *usec) {
...
+
+int parse_timestamp(const char *t, usec_t *usec) {
+        char *last_space, *timezone = NULL;
+        ParseTimestampResult *shared, tmp;
+        int r;
+        pid_t pid;
+
+        last_space = strrchr(t, ' ');
+
+        if (last_space != NULL) {
```
Comment (keszybz) [inline src/basic/time-util.c]: Hmm, you have the exact same code in `calendar_spec_next_usec` above. This needs to be extracted into a helper function, not repeated.

File: src/basic/calendarspec.c (hunk 20 above)
Comment (zerkms) [reply src/basic/calendarspec.c]: @keszybz `strrchr` does not return a `const char *`, but `char *`.

File: src/basic/time-util.c (hunk 23 above)
Comment (zerkms) [reply src/basic/time-util.c]: @keszybz I was not sure whether it makes sense to have it implemented first, then refactored as a part of the next PR.

Comment (zerkms) [reply src/basic/time-util.c]: @keszybz they are vaguely similar. It duplicates statements, not the knowledge. (so it's not a DRY violation)

File: src/basic/calendarspec.c (hunk 20 above)
Comment (keszybz) [reply src/basic/calendarspec.c]: The cast from `char *` to `const char*` is implicit.

File: src/basic/time-util.c (hunk 23 above)
Comment (keszybz) [reply src/basic/time-util.c]: They are sufficiently similar to unify the common part: lines 1362–1400.

File: src/basic/calendarspec.c (hunk 20 above)
Comment (zerkms) [reply src/basic/calendarspec.c]: Right, if both are `const` it's fine, otherwice `gcc` emits a warning. Fixed as well.

File: src/basic/time-util.c (hunk 23 above)
Comment (zerkms) [reply src/basic/time-util.c]: Not sure how nice it would look like though: they are different in what main "implementation" they call. They are still 2 different implementations that happen to have some parts similar (since they do similar work). This similarity is accidental, and there is no relationship between them to expect them to evolve identically.

I will try to extract it to a shared common function, but its use would always be limited by these 2 cases only.

I personally would not do that, but let's try and see.

Comment (zerkms) [reply src/basic/time-util.c]: @keszybz I spent some time trying and I don't see how it's possible to generalise it in a way that does not make it 10 times more complicated spaghetti of (not reusable) functions and macroses.

If you have any particular idea - please share it.

Comment (keszybz) [reply src/basic/time-util.c]: OK, so please push the best version you have without this change, and we'll go from there.

Comment (zerkms) [reply src/basic/time-util.c]: @keszybz what currently is commited contains all the changes requested from you and Lennart.

Do I need to squash/rebase it again?

**Comment by keszybz** [comment]
I implemented the change to unify both functions locally, and you were right, it's more trouble than it's worth. So let's not do that ;)

I squashed your commits and pushed two fixup on top (that do no functional change, except adding a logging statement).

Unfortunately, your change breaks the tests, when run in non-UTC/non-CEST locale:
```console
$ SYSTEMD_LOG_LEVEL=err TZ=UTC build2/test-date
$ SYSTEMD_LOG_LEVEL=err TZ=CEST build2/test-date
$ SYSTEMD_LOG_LEVEL=err TZ=CET build2/test-date
round-trip failed: "Sun 2012-12-30 18:42:00.000000 CET" → "Sun 2012-12-30 17:42:00.000000 CET"
Assertion 'q == t' failed at ../src/test/test-date.c:42, function test_should_pass(). Aborting.
[3]    23267 abort (core dumped)  SYSTEMD_LOG_LEVEL=err TZ=CET build2/test-date
```

**Comment by zerkms** [comment]
~There is a chance it's tests that need some redesign, not the implementation itself, because:~

```
$ TZ=CET date
Sun Sep 17 12:32:22 CEST 2017
$ TZ=CEST date
Sun Sep 17 10:32:23 CEST 2017
$ TZ=FOOBAR date
Sun Sep 17 12:32:25 FOOBAR 2017
```

I'm not sure it's correct to set `CEST` via `TZ`, since it's not even a real timezone name. Apparently, for every unknown timezone it just accepts it as UTC.

It needs some more work and I'll make a proper research tomorrow (it's time to sleep here).

PS: yep, quick debugging reveals some problems with implementation. I couldn't believe introducing something as complicated as timezones might be that simple...

**Comment by zerkms** [comment]
@keszybz 

Couldn't resist fixing it.

The problem was with the `tm.tm_isdst` that is set to the current environment value: either DST or not. While the current state is not relevant to the state in the desired date.

Hence - it should be reset so that the `mktime_or_timegm` could normalise it later.

**Comment by keszybz** [comment]
I force-pushed once more: no changes to the code, but I added your last comment to the commit message.

**Comment by zerkms** [comment]
@keszybz is it alright that the latest `master` commit failed to build: https://www.dropbox.com/s/80fn1rxxv2u842l/Screenshot%202017-09-18%2007.57.47.png?dl=0

I checked to the CI but not sure where to even look at to see the error message or anything.

**Comment by keszybz** [comment]
Not sure. I kicked off a rebuild.

**Comment by zerkms** [comment]
What's the correct way to see the output of the commands on semaphoreci?

https://semaphoreci.com/systemd/systemd/branches/master/builds/3493 - I'm looking at the current build in progress and none of the commands have any output :-S

**Comment by keszybz** [comment]
Yeah, it seems that something is borked in semaphore. Normally, you can see the output of all commands (until the one that failed).

**Comment by ervinb** [comment]
@keszybz We've identified the issue, and the instance is under investigation. Your builds should be back to normal. Get in touch on support if you notice anything strange. Thanks for your understanding!

**Comment by japm48** [comment]
Just a question: will this respect DST?
Example: the expected behaviour for `daily Europe/Paris` is a 24 hour span, in general, between two consecutive triggers, except the two DST transition days, in which it should be 23 h and 25 h (leap seconds aside).
As I'm coordinating with "local" events I cannot use UTC for this.

**Comment by zerkms** [comment]
> Example: the expected behaviour for daily Europe/Paris is a 24 hour span

The expected behaviour for `daily Europe/Paris` is "every midnight in Paris". It might be 24 hours, 23 hours, 25 hours or 24 hours and 1 second.

> Just a question: will this respect DST?

So, yep. It will fire every midnight according the Paris time, taking into account the current time in Paris.

**Comment by japm48** [comment]
@zerkms Thank you. I already knew the "default is midnight" (= 00:00), just wanted to make sure it won't do weird unexpected things; well, DST is weird and stupid by itself (IMHO) but nothing can be done about that...

Besides, I'm not using 2 or 3 PM (DST change times) so I won't hit something like issue #5595. When that issue is solved, it would be sensible to add tests for those particular edge cases (?). At "DST change days" it may trigger zero, one or two times, depending on the chosen expected behaviour; IMO, it should be always just once but I understand the other perspectives.

I guess timing is a minefield of controversial edge cases.

**Comment by zerkms** [comment]
> IMO, it should be always just once but I understand the other perspectives.

It's hard to guess what exactly a user tried to express: the current implementation is the most dumb and makes no assumption at all. So it's up to a user to avoid any unambiguity :-)

> When that issue is solved

I'm not sure it's a good idea to solve it though.

> IMO, it should be always just once but I understand the other perspectives.

Given something is scheduled to run monthly at 31st, do you expect it to run in February, April and June?

The same for: `February 29` - if we follow what you suggested - it would be run every year. But what if the original intention was "every leap year"?
